server/internal/application/user/user.go

```diff
@@ -0,0 +1,62 @@
+package user
+
+import (
+	"errors"
+	"time"
+)
+
+var (
+	ErrNameNotValid = errors.New("pippo is not valid")
+)
+
+type User struct {
+	Username  string
+	Password  string
+	Name      string
+	Surname   string
+	Email     string
+	Role      string
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+type UserCreateDTO struct {
+	ID        string
+	Username  string
+	Password  string
+	Name      string
+	Surname   string
+	Email     string
+	Role      string
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+type UserUpdateDTO struct {
+	ID       string
+	Username *string
+	Password *string
+	Name     *string
+	Surname  *string
+	Email    *string
+	Role     *string
+}
+
+type UserRetrieveDTO struct {
+	ID        string
+	Username  string
+	Password  string
+	Name      string
+	Surname   string
+	Email     string
+	Role      string
+	CreatedAt time.Time
+	UpdatedAt time.Time
+}
+
+type UserRepository interface {
+	Register(u UserCreateDTO) error
+	Update(u UserUpdateDTO) error
+	FindByUsername(username string) (UserRetrieveDTO, error)
+	FindByEmail(email string) (UserRetrieveDTO, error)
+}
```

Comment:
Naming nitpicking: I like to name repository interfaces so that they "read" well when used. Consider the following code:

```go
type UserRepository interface {
	FindByUsername(username string) (*User, error)
}

userRepo := infrastructure.NewPgUserRepo()
user, err := userRepo.FindByUsername(username)
```

It's long, verbose, and knowing that it's a repository doesn't really add any value. Consider this form instead:


```go
type Users interface {
	ByUsername(username string) (*User, error)
}

users := infrastructure.NewPgUsers()
user, err := users.ByUsername(username)
```
It's lighter, less concerned with technical code terms, and it reads well in English ("lookup users by username")